Autocalculate DOAS inputs in Sizing:Zone. Note I haven't figured out …

wfbuhl opened the pull request:
## Pull request overview

Please change this line to a description of the pull request, with useful supporting information including whether it is a new feature, or fixes a defect, a cross reference to any defects addressed in this PR, the type of changes to be made, and whether diffs are expected/justified based on this change.
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [ ] If defect, results of running current develop vs this branch should exhibit the fix
- [ ] CI status: all green or justified
- [ ] Performance: CI Linux results include performance check -- verify this
- [ ] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] If transition, add rules to spreadsheet
  - [ ] If transition, add transition source
  - [ ] If transition, update idfs
- [ ] If new idf included, locally check the err file and other outputs
- [ ] Documentation changes in place
- [ ] ExpandObjects changes??
- [ ] If output changes, including tabular output structure, add to output rules file for interfaces

…how/where to put this out onto the eio file.

wfbuhl committed:
Autocalculate DOAS inputs in Sizing:Zone. Note I haven't figured out how/where to put this out onto the eio file.

wfbuhl committed:
Added eio output for Sizing:Zone DOAS inputs when "Account for Dedicated Outside Air System" = Yes. No SQL output for this yet. Otherwise this is ready to go.

wfbuhl committed:
Added unit test

Myoldmopar commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -5455,6 +5459,164 @@ namespace ZoneEquipmentManager {
+		// na
+
+		// SUBROUTINE LOCAL VARIABLE DECLARATIONS:
+		static bool MyOneTimeFlag( true );
```
@wfbuhl So these static variables inside functions cause issues with unit tests.  Basically, in unit tests, there is no way to go back into the function and reset the value, so the effects can fall downstream to the _next_ unit test that is run in the same session.

I would:
- rename this variable to something clear for this application, 
- put it in the anonymous namespace at the [top of this file](https://github.com/NREL/EnergyPlus/blob/94441801dcd7780ba27759c138f50bcc360e1792/src/EnergyPlus/ZoneEquipmentManager.cc#L191) (Fortran-analogy: make it a module variable), and
- make sure the value gets reset inside the [clear_state function](https://github.com/NREL/EnergyPlus/blob/94441801dcd7780ba27759c138f50bcc360e1792/src/EnergyPlus/ZoneEquipmentManager.cc#L215) already defined for this namespace so the value is reset for any following unit tests that get run.

wfbuhl replied:
Good ideas. Of course this is just a modified copy of ReportSizingOutput.
So similar issues are scattered through the code.

On Tue, Mar 22, 2016 at 8:27 AM, Edwin Lee notifications@github.com wrote:

> In src/EnergyPlus/ZoneEquipmentManager.cc
> https://github.com/NREL/EnergyPlus/pull/5557#discussion_r57006747:
> 
> > -       using DataStringGlobals::VerString;
> > -       using General::RoundSigDigits;
> >   +
> > -       // Locals
> > -       // SUBROUTINE ARGUMENT DEFINITIONS:
> >   +
> > -       // SUBROUTINE PARAMETER DEFINITIONS:
> >   +
> > -       // INTERFACE BLOCK SPECIFICATIONS
> > -       // na
> >   +
> > -       // DERIVED TYPE DEFINITIONS
> > -       // na
> >   +
> > -       // SUBROUTINE LOCAL VARIABLE DECLARATIONS:
> > -       static bool MyOneTimeFlag( true );
> 
> @wfbuhl https://github.com/wfbuhl So these static variables inside
> functions cause issues with unit tests. Basically, in unit tests, there is
> no way to go back into the function and reset the value, so the effects can
> fall downstream to the _next_ unit test that is run in the same session.
> 
> I would:
> - rename this variable to something clear for this application,
> - put it in the anonymous namespace at the top of this file
>   https://github.com/NREL/EnergyPlus/blob/94441801dcd7780ba27759c138f50bcc360e1792/src/EnergyPlus/ZoneEquipmentManager.cc#L191
>   (Fortran-analogy: make it a module variable), and
> - make sure the value gets reset inside the clear_state function
>   https://github.com/NREL/EnergyPlus/blob/94441801dcd7780ba27759c138f50bcc360e1792/src/EnergyPlus/ZoneEquipmentManager.cc#L215
>   already defined for this namespace so the value is reset for any following
>   unit tests that get run.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5557/files/94441801dcd7780ba27759c138f50bcc360e1792#r57006747

Myoldmopar commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -5455,6 +5459,164 @@ namespace ZoneEquipmentManager {
 		MassConservation(ZoneNum).MixingSourceMassFlowRate = ZoneSourceMassFlowRate;
 	}
 
+	void
```
This is a nice, packaged up, routine with a clear purpose.  Easy to unit test.  This is good.  I have only three comments, all minor, and more just comments for future coding than this branch.
- For some reason, the function name indentation is 2 tabs deep, not sure why.  If you happen to go back into the code for any changes, just take a tab out and it will fall in line with the rest of the code.
- I would take out the unused portions of the function comment header.  It doesn't hurt that they are there, and since you have included a note in the REFERENCES section, I'd leave that.  Just the unused na ones dont need to be included.
- This would have been ripe for a very tiny move toward object oriented coding.  Here's how it would have been done:

`ZoneSizingInput` is an array of type `ZoneSizingInputData`.  If you go over to that struct definition, you could add a function declaration like this:

```
struct ZoneSizingInputData {
  // lots of stuff
  // variables, etc.
  void sizeDOASControlStrategy();
  // other stuff
}
```

Now there would be a function called sizeDOASControlStrategy that is a _member_ of the struct.  To call this function, you would then loop over the instances and call that method:

```
for ( auto & thisZoneData : ZoneSizingInput ) {
  thisZoneData.sizeDOASControlStrategy();
}
```

So far so good.  Then inside `sizeDOASControlStrategy`, you would just operate on the current instance:

```
void ZoneSizingInputData::sizeDOASControlStrategy() {
  // comments
  if ( this->AccountForDOAS ) {
    if ( this->DOASControlStrategy == DOASNeutralSup ) {
      if ( this->DOASLowSetpoint == AutoSize && this->DOASHighSetpoint == AutoSize ) {
        this->DOASLowSetpoint = 21.1;
        this->DOASHighSetpoint = 23.9;
      }
  // etc...
```

Inside that function, you are always operating on the "current zone sizing input", or "THIS zone sizing input".  That's the purpose of the `this->`, which is actually optional.  

This would be a small change, but would be one step closer to objectification, and even easier to unit test, in my opinion.  And it's also a form of "gateway" into OO since it is such an isolated change.

OK, that is all, thanks.

wfbuhl replied:
This is modest enough that it can squeeze into my unused brain cells.

On Tue, Mar 22, 2016 at 8:40 AM, Edwin Lee notifications@github.com wrote:

> In src/EnergyPlus/ZoneEquipmentManager.cc
> https://github.com/NREL/EnergyPlus/pull/5557#discussion_r57009446:
> 
> > @@ -5455,6 +5459,164 @@ namespace ZoneEquipmentManager {
> >         MassConservation(ZoneNum).MixingSourceMassFlowRate = ZoneSourceMassFlowRate;
> >     }
> > -   void
> 
> This is a nice, packaged up, routine with a clear purpose. Easy to unit
> test. This is good. I have only three comments, all minor, and more just
> comments for future coding than this branch.
> - For some reason, the function name indentation is 2 tabs deep, not
>   sure why. If you happen to go back into the code for any changes, just take
>   a tab out and it will fall in line with the rest of the code.
> - I would take out the unused portions of the function comment header.
>   It doesn't hurt that they are there, and since you have included a note in
>   the REFERENCES section, I'd leave that. Just the unused na ones dont need
>   to be included.
> - This would have been ripe for a very tiny move toward object
>   oriented coding. Here's how it would have been done:
> 
> ZoneSizingInput is an array of type ZoneSizingInputData. If you go over
> to that struct definition, you could add a function declaration like this:
> 
> struct ZoneSizingInputData {
>   // lots of stuff
>   // variables, etc.
>   void sizeDOASControlStrategy();
>   // other stuff
> }
> 
> Now there would be a function called sizeDOASControlStrategy that is a
> _member_ of the struct. To call this function, you would then loop over
> the instances and call that method:
> 
> for ( auto & thisZoneData : ZoneSizingInput ) {
>   thisZoneData.sizeDOASControlStrategy();
> }
> 
> So far so good. Then inside sizeDOASControlStrategy, you would just
> operate on the current instance:
> 
> void ZoneSizingInputData::sizeDOASControlStrategy() {
>   // comments
>   if ( this->AccountForDOAS ) {
>     if ( this->DOASControlStrategy == DOASNeutralSup ) {
>       if ( this->DOASLowSetpoint == AutoSize && this->DOASHighSetpoint == AutoSize ) {
>         this->DOASLowSetpoint = 21.1;
>         this->DOASHighSetpoint = 23.9;
>       }
>   // etc...
> 
> Inside that function, you are always operating on the "current zone sizing
> input", or "THIS zone sizing input". That's the purpose of the this->,
> which is actually optional.
> 
> This would be a small change, but would be one step closer to
> objectification, and even easier to unit test, in my opinion. And it's also
> a form of "gateway" into OO since it is such an isolated change.
> 
> OK, that is all, thanks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5557/files/94441801dcd7780ba27759c138f50bcc360e1792#r57009446

Myoldmopar commented:
So do you _anticipate_ doing SQL output for this?  Or are you saying this will have to wait for later and you want to wrap this up as-is?

Myoldmopar commented:
(Saw your email @wfbuhl, will communicate there)

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into Issue5491

Myoldmopar committed:
Fix minor indentation issues

Myoldmopar committed:
Move static var to anonymous namespace and add to clear_state

Myoldmopar commented:
OK, I merged in develop, cleaned up the indentation issues, and moved that variable out of the function and added it to the clear_state function.  Since we won't be holding this up for the SQL output portion, this should be good to go as long as CI comes back happy.

Myoldmopar commented:
Everything looks hunky dory; the only diffs being the eio line additions, so this should drop right in.

wfbuhl commented:
Make a new issue on verification of sizing inputs. This is getting somewhat
important since sizing input is getting more complicated. Should come out
on tabular and SQL. From my quick scan of the SQL routine, there seems to
be some of it hard-coded and some that is more flexible. The thing that
baffled me was how to assign labels or tags to new data. I'm sure I could
figure it out if given a little time. Tabular I've done before. But it
didn't seem worth the effort to just do this for DOAS sizing. The whole
shebang needs to be output to tabular and SQL Also I haven't looked into
how to test SQL.

On Tue, Mar 22, 2016 at 8:41 AM, Edwin Lee notifications@github.com wrote:

> So do you _anticipate_ doing SQL output for this? Or are you saying this
> will have to wait for later and you want to wrap this up as-is?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5557#issuecomment-199871821

Myoldmopar commented:
> Make a new issue on verification of sizing inputs.

Are you saying you are doing it?  Or was that a command?  Because if it was a command, you're going to need to prefix with `sudo`.

wfbuhl commented:
Should only change the one eio file.

On Tue, Mar 22, 2016 at 9:04 AM, Edwin Lee notifications@github.com wrote:

> OK, I merged in develop, cleaned up the indentation issues, and moved that
> variable out of the function and added it to the clear_state function.
> Since we won't be holding this up for the SQL output portion, this should
> be good to go as long as CI comes back happy.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5557#issuecomment-199882207

wfbuhl commented:
Well now I don't have anything to do today.

On Tue, Mar 22, 2016 at 9:04 AM, Edwin Lee notifications@github.com wrote:

> OK, I merged in develop, cleaned up the indentation issues, and moved that
> variable out of the function and added it to the clear_state function.
> Since we won't be holding this up for the SQL output portion, this should
> be good to go as long as CI comes back happy.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5557#issuecomment-199882207

wfbuhl commented:
See previous.

On Tue, Mar 22, 2016 at 8:46 AM, Edwin Lee notifications@github.com wrote:

> (Saw your email @wfbuhl https://github.com/wfbuhl, will communicate
> there)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5557#issuecomment-199873870

wfbuhl commented:
I will do it since my day has suddenly opened up.

On Tue, Mar 22, 2016 at 10:47 AM, Edwin Lee notifications@github.com
wrote:

> Make a new issue on verification of sizing inputs.
> 
> Are you saying you are doing it? Or was that a command? Because if it was
> a command, you're going to need to prefix with sudo.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5557#issuecomment-199932866
